core/services/workflows/engine.go

```diff
@@ -12,49 +15,91 @@ import (
 )
 
 const (
-	mockedWorkflowID = "ef7c8168-f4d1-422f-a4b2-8ce0a1075f0a"
-	mockedTriggerID  = "bd727a82-5cac-4071-be62-0152dd9adb0f"
+	mockedWorkflowID  = "aaaaaaaa-f4d1-422f-a4b2-8ce0a1075f0a"
+	mockedExecutionID = "bbbbbbbb-f4d1-422f-a4b2-8ce0a1075f0a"
+	mockedTriggerID   = "cccccccc-5cac-4071-be62-0152dd9adb0f"
 )
 
 type Engine struct {
 	services.StateMachine
-	logger     logger.Logger
-	registry   types.CapabilitiesRegistry
-	trigger    capabilities.TriggerCapability
-	consensus  capabilities.ConsensusCapability
-	target     capabilities.TargetCapability
-	callbackCh chan capabilities.CapabilityResponse
-	cancel     func()
+	logger          logger.Logger
+	registry        types.CapabilitiesRegistry
+	triggerType     string
+	triggerConfig   *values.Map
+	trigger         capabilities.TriggerCapability
+	consensusType   string
+	consensusConfig *values.Map
+	consensus       capabilities.ConsensusCapability
+	targetType      string
+	targetConfig    *values.Map
+	target          capabilities.TargetCapability
+	callbackCh      chan capabilities.CapabilityResponse
+	cancel          func()
 }
 
 func (e *Engine) Start(ctx context.Context) error {
 	return e.StartOnce("Engine", func() error {
-		err := e.registerTrigger(ctx)
-		if err != nil {
-			return err
-		}
-
 		// create a new context, since the one passed in via Start is short-lived.
 		ctx, cancel := context.WithCancel(context.Background())
 		e.cancel = cancel
-		go e.loop(ctx)
+		go e.init(ctx)
+		go e.triggerHandlerLoop(ctx)
 		return nil
 	})
 }
 
-func (e *Engine) registerTrigger(ctx context.Context) error {
-	triggerConf, err := values.NewMap(
-		map[string]any{
-			"feedlist": []any{
-				// ETHUSD, LINKUSD, USDBTC
-				123, 456, 789,
-			},
+func (e *Engine) init(ctx context.Context) {
+	retrySec := 5
+	ticker := time.NewTicker(time.Duration(retrySec) * time.Second)
+	defer ticker.Stop()
+	var err error
+LOOP:
+	for {
```

Comment:
Rather than a retry loop, why not a oneshot channel we can `close` to signal after all the registration is done?

Reply:
Who would signal that? Currently our registrations are done all over the place :( In the future we can improve that but even then we could have remote capabilities being registered at arbitrary times.